### PR TITLE
Use default Plum dispatcher object

### PR DESCRIPTION
**Pull request by PhilipVinc**
Let's just use the default dispatcher so we don't have to always import the same from a central point, and it will make easier for other packages to extend netket in the future.

### CODE REVIEW

**Comment by femtobit** [comment]
I am not sure about that. The function `product` has a somewhat general name. The only case where one would ever want to extend that is when adding simplification rules for custom element types (as with the grid-specific translations and rotations in `grid.py`). In that case, importing `semigroup.dispatch` is straightforward and seems very logical to me.

Having a single `dispatch` for the group elements makes sure this is done predictably in one place.

I wouldn't mind using the global dispatcher for netket functions that deal, e.g., with different array types or similar but our "group algebra" seems to be a special case to me.

**Comment by femtobit** [comment]
Since all elements inherit from `Callable`, I think it's possible to accidentally add an overload that influences the group operations if it also matches `product(Callable, <something>)`. Maybe I'm too careful, bvut OTOH I don't really see a downside (or barrier to extensibility) when compartmentalizing the `distpatch` in this case.

**Comment by PhilipVinc** [comment]
So there are three possibile things that can happen:
 - An user define a custom element type `Custom`, and overloads `product(: Custom, : Custom )` . Custom does not exist before so this method will never hit anything defined in base netket. This is fine
 - An user defines a custom element type `Custom` and overloads `product(: Custom, : netket.Element)`, where Element is an Element defined in base netket. As the user has implemented custom himselfs, he knows how product should behave and that method will work correctly. Likewise, as `Custom` is not used in base, this method he defined will never affect base netket code.
 - An user defines `product(: netket.Element, : Callable )`. Since all types appearing in this dispatch signature already exist and have not been defined by the user, this code `might` affect and change the behaviour of base netket. This is known as [Type Piracy](https://docs.julialang.org/en/v1/manual/style-guide/#Avoid-type-piracy). If you commit it, you know that you `might` break stuff. At the same time, fobidding/preventing it prevents people from playing with the code and modifying it`s behaviour for countless of valid reasons, so i am against it.

In short: If you are defining dispatch signatures with only types you do not _own_, you are committing piracy and you should be aware of it. 
In Julia packages do not commit type piracy, yet everybody uses the same */-+, prod, sum. 

--

I think we only need one dispatch object for the whole of netket.
If you are afraid other packages define `product` and create problems for us... We might define our own dispatch object but I still think that this is kinda-ugly, though acceptable.

**Comment by PhilipVinc** [comment]
Also, since you correctly defined
```python
@dispatch(Element, Element)
def product(a: Element, b: Element):
    return Composite(a, b)
```

this means that one cannot erroneously override our logic by defining an arbitrary dispatch signature.
He must explicitly insert a subtype of Element in the signature.

**Review by femtobit — APPROVED** [review]
Alright, I'm convinced.
